### PR TITLE
CASSANDRA-20744: Add a simulator test using the Harry AST logic

### PR DESCRIPTION
**Pull request by dcapwell**

### CODE REVIEW

File: src/java/org/apache/cassandra/cql3/statements/BatchStatement.java [hunk 1]
```diff
@@ -357,7 +357,7 @@ public List<? extends IMutation> getMutations(ClientState state,
         }
         // local is either executeWithoutConditions modifying a virtual table (doesn't support txns) or executeLocal
         // which is called by test or internal things that are bypassing distributed system modification/checks
-        return collector.toMutations(state, local ? PotentialTxnConflicts.ALLOW : PotentialTxnConflicts.DISALLOW);
+        return collector.toMutations(state, local ? PotentialTxnConflicts.ALLOW : PotentialTxnConflicts.DISALLOW, attrs.isTimestampSet());
```
Comment (dcapwell) [inline src/java/org/apache/cassandra/cql3/statements/BatchStatement.java]: this is dead code atm, i pass it in but its not used.

```
BEGIN BATCH
   INSERT INTO tbl(pk, ck, v) VALUES (0, 0, 0) USING TIMESTAMP 42;
   INSERT INTO tbl(pk, ck, v) VALUES (0, 1, 1);
```

in this example the first mutation is using the user timestamp, but the second is using the server timestamp; they are the same `Mutation` object as they share the same PK!

File: src/java/org/apache/cassandra/db/ColumnFamilyStore.java [hunk 2]
```diff
@@ -2228,6 +2228,12 @@ public void forceMajorCompaction(boolean splitOutput, int permittedParallelism)
         CompactionManager.instance.performMaximal(this, splitOutput, permittedParallelism);
     }
 
+    @VisibleForTesting
+    public List<Future<?>> submitMajorCompaction(boolean splitOutput, int permittedParallelism)
```
Comment (dcapwell) [inline src/java/org/apache/cassandra/db/ColumnFamilyStore.java:2233]: i mostly added this for Caleb as he wanted a AST based test that had async compact... but turns out Simulator will still make the compaction blocking so this change doesn't *really* matter at the moment.  I can revert and using the blocking logic as it doesn't change the test, and leave async compaction for caleb in a followup patch (which is needed anyways)

File: src/java/org/apache/cassandra/db/Mutation.java [hunk 3]
```diff
@@ -552,17 +570,18 @@ private Serialization serialization(Mutation mutation, int version)
...
         {
             Map<TableId, PartitionUpdate> modifications = mutation.modifications;
 
             if (version >= VERSION_51)
             {
                 int flags = 0;
                 flags |= potentialTxnConflictsFlag(mutation.potentialTxnConflicts);
-                out.write(flags);
+                flags |= userTimestampFlag(mutation.userTimestamp);
+                out.writeUnsignedVInt32(flags);
```
Comment (dcapwell) [inline src/java/org/apache/cassandra/db/Mutation.java]: rather than write 4 bytes, write 1.

File: src/java/org/apache/cassandra/db/compaction/CompactionManager.java [hunk 4]
```diff
@@ -1141,6 +1141,11 @@ public void performMaximal(final ColumnFamilyStore cfStore, boolean splitOutput,
         FBUtilities.waitOnFutures(submitMaximal(cfStore, getDefaultGcBefore(cfStore, FBUtilities.nowInSeconds()), splitOutput, permittedParallelism));
     }
 
+    public List<Future<?>> submitMaximal(final ColumnFamilyStore cfStore, boolean splitOutput, int permittedParallelism)
```
Comment (dcapwell) [inline src/java/org/apache/cassandra/db/compaction/CompactionManager.java:1144]: same comment as above, this change isn't actually required by this patch, it was there to make it easier for Caleb to do follow up work.  Can always drop if desired.

File: src/java/org/apache/cassandra/service/StorageProxy.java [hunk 5]
```diff
@@ -1248,20 +1249,22 @@ public static void mutateWithTriggers(List<? extends IMutation> mutations,
...
-    public static void dispatchMutationsWithRetryOnDifferentSystem(List<? extends IMutation> mutations, ConsistencyLevel consistencyLevel, Dispatcher.RequestTime requestTime)
+    public static void dispatchMutationsWithRetryOnDifferentSystem(List<? extends IMutation> mutations, ConsistencyLevel consistencyLevel, Dispatcher.RequestTime requestTime, boolean preserveTimestamps)
     {
         while (true)
         {
             ClusterMetadata cm = ClusterMetadata.current();
             try
             {
-                SplitMutations splitMutations = splitMutationsIntoAccordAndNormal(cm, (List<IMutation>)mutations);
+                SplitMutations<?> splitMutations = splitMutationsIntoAccordAndNormal(cm, (List<IMutation>)mutations);
```
Comment (dcapwell) [inline src/java/org/apache/cassandra/service/StorageProxy.java:1266]: just fixing intellij warning

File: src/java/org/apache/cassandra/service/accord/txn/TxnWrite.java [hunk 6]
```diff
@@ -475,6 +489,30 @@ public long estimatedSizeOnHeap()
...
 
+    private static int getFlags(Update write)
+    {
+        return getFlags(() -> write.userTimestamp);
+    }
+
+    private static int getFlags(Fragment write)
+    {
+        return getFlags(() -> write.userTimestamp);
+    }
```
Comment (dcapwell) [inline src/java/org/apache/cassandra/service/accord/txn/TxnWrite.java]: its done this way to make it testable

File: test/distributed/org/apache/cassandra/distributed/test/cql3/AccordInteropMultiNodeTableWalkBase.java [hunk 7]
```diff
@@ -83,9 +85,27 @@ protected State createState(RandomSource rs, Cluster cluster)
...
+            allowUsingTimestamp = rs.nextBoolean();
+        }
+
+        @Override
+        protected void createTable(TableMetadata metadata)
+        {
+            super.createTable(metadata);
+            Epoch maxEpoch = ClusterUtils.maxEpoch(cluster);
+            ClusterUtils.waitForCMSToQuiesce(cluster, maxEpoch);
+            ClusterUtils.awaitAccordEpochReady(cluster, maxEpoch.getEpoch());
```
Comment (dcapwell) [inline test/distributed/org/apache/cassandra/distributed/test/cql3/AccordInteropMultiNodeTableWalkBase.java:102]: race condition bug making these tests flakey.  We would start work *before* it was ready in accord.

File: test/distributed/org/apache/cassandra/distributed/test/cql3/AccordInteropMultiNodeTableWalkBase.java [hunk 8]
```diff
@@ -83,9 +85,27 @@ protected State createState(RandomSource rs, Cluster cluster)
 
     public class AccordInteropMultiNodeState extends MultiNodeState
     {
+        private final boolean allowUsingTimestamp;
+
         public AccordInteropMultiNodeState(RandomSource rs, Cluster cluster)
         {
             super(rs, cluster);
+            allowUsingTimestamp = rs.nextBoolean();
```
Comment (dcapwell) [inline test/distributed/org/apache/cassandra/distributed/test/cql3/AccordInteropMultiNodeTableWalkBase.java:93]: run the tests with and without `USING TIMESTAMP`.

This impacts my other branch (can't merge as its finding bugs still needing to be fixed) that adds `BEGIN TRANSACTION` support... i just need to block using `BEGIN TRANSACTION` when this is `true`

File: test/harry/main/org/apache/cassandra/harry/model/ASTSingleTableModel.java [hunk 9]
```diff
@@ -83,10 +84,26 @@
 import org.apache.cassandra.utils.Pair;
 
 import static org.apache.cassandra.cql3.ast.Elements.symbols;
+import static org.apache.cassandra.harry.MagicConstants.NO_TIMESTAMP;
 import static org.apache.cassandra.harry.model.BytesPartitionState.asCQL;
 
 public class ASTSingleTableModel
 {
+    private enum CasResponse
```
Comment (dcapwell) [inline test/harry/main/org/apache/cassandra/harry/model/ASTSingleTableModel.java:92]: this is dead code in this patch... its for the `validateCasAppliedOnly` method which is used in my other branch... i can revert, it was just easier to pull forward all AST changes rather than cherry picking...


The issue is that BATCH + Paxos is... annoying... so can't validate the result schema yet.  The other issue is that paxos v1/v2 on accord doesn't have the right schema, so when adding proper accord tests they fail almost always... so have to limit to just the `applied` column rather than the full schema...

The bug is paxos v1/v2 *leak* implementation details to the user, and since Accord does reads differently the leaked details don't match.  This is a bug that needs to be fixed, but i was trying to get actual coverage of paxos rather than focusing on fixing the failed applied schema issue...

File: test/harry/main/org/apache/cassandra/harry/model/ASTSingleTableModel.java [hunk 10]
```diff
@@ -670,19 +781,36 @@ else if (condition.getClass() == Conditional.Where.class)
...
+                case cas:
+                {
+                    // If anything is null avoid doing the test, but there is a special case where this returns true... both sides are null!
+                    // This logic isn't consistent with other parts of the database and is local to CAS IF clause
+                    // see ML@Inconsistent null handling between WHERE and IF clauses
+                    if (lhs == null || rhs == null)
+                        return lhs == rhs;
+                }
+                break;
+                case accord:
```
Comment (dcapwell) [inline test/harry/main/org/apache/cassandra/harry/model/ASTSingleTableModel.java:795]: not implemented in this patch, its from my other patch trying to add `BEGIN TRANSACTION` support.

That ticket found this issue, which caused me to create a specialized simulator test to reproduce to validate changes...

File: test/simulator/main/org/apache/cassandra/simulator/ActionSchedule.java [hunk 11]
```diff
@@ -73,7 +73,65 @@ public class ActionSchedule implements CloseableIterator<Object>, LongConsumer
 {
     private static final Logger logger = LoggerFactory.getLogger(ActionList.class);
 
-    public enum Mode { TIME_LIMITED, STREAM_LIMITED, TIME_AND_STREAM_LIMITED, FINITE, UNLIMITED }
+    public enum Mode
+    {
```
Comment (dcapwell) [inline test/simulator/main/org/apache/cassandra/simulator/ActionSchedule.java:77]: this was very painful for me to figure out what mode to use... so figured adding docs would help the next person.

The docs were started by Claude, but refined by me.

File: test/simulator/main/org/apache/cassandra/simulator/Actions.java [hunk 12]
```diff
@@ -33,6 +33,24 @@
 
 public class Actions
 {
+    public static class EmptyAction extends Action
```
Comment (dcapwell) [inline test/simulator/main/org/apache/cassandra/simulator/Actions.java]: i need a way to ignore empty actions so i can pick a better action to take.  The current behavior has no clear way to know its empty

File: test/simulator/main/org/apache/cassandra/simulator/ClusterSimulation.java [hunk 13]
```diff
@@ -141,6 +141,21 @@ public interface SchedulerFactory
...
 
+    public interface FutureActionSchedulerFactory
+    {
+        FutureActionScheduler create(int nodeCount, SimulatedTime time, RandomSource random);
+    }
+
+    public interface PerVerbFutureActionSchedulersFactory
+    {
+        Map<Verb, FutureActionScheduler> create(int nodeCount, SimulatedTime time, RandomSource random);
+    }
```
Comment (dcapwell) [inline test/simulator/main/org/apache/cassandra/simulator/ClusterSimulation.java:152]: this patch is single node so this really doesn't matter... my first simulator POC was multi DC where this mattered more.

The reason this patch is single node was that there was simulator bugs and Benedict told me that Ariel was working on several bug fixes, so it was easier to make the test single node for now.

The thinking is that once Ariel fixes the open issues, we can add a *new* test that turns the AST test into a multi node test (just like how the jvm-dtest tests work)

File: test/simulator/main/org/apache/cassandra/simulator/ClusterSimulation.java [hunk 14]
```diff
@@ -141,6 +141,21 @@ public interface SchedulerFactory
...
+
+    public interface TimeFactory
+    {
+        SimulatedTime create(RandomSource random, int numOfNodes);
+    }
+
```
Comment (dcapwell) [inline test/simulator/main/org/apache/cassandra/simulator/ClusterSimulation.java]: going to revert this, this was an idea benedict and I played around with to try to repo the underline issue, but this was not actually needed to repo.

Comment (dcapwell) [reply test/simulator/main/org/apache/cassandra/simulator/ClusterSimulation.java]: removed

File: test/simulator/main/org/apache/cassandra/simulator/ClusterSimulation.java [hunk 15]
```diff
@@ -716,7 +744,7 @@ public ClusterSimulation(RandomSource random, long seed, int uniqueNum,
             {
                 int numInDc = (numOfNodes / numOfDcs) + (numOfNodes % numOfDcs > i ? 1 : 0);
                 numInDcs[i] = numInDc;
-                minRf[i] = 3;
+                minRf[i] = min(numInDc, 3);
```
Comment (dcapwell) [inline test/simulator/main/org/apache/cassandra/simulator/ClusterSimulation.java:747]: bug with single node clusters.

File: test/simulator/main/org/apache/cassandra/simulator/systems/InterceptibleThreadFactory.java [hunk 16]
```diff
@@ -58,7 +58,9 @@ public InterceptibleThread newThread(Runnable runnable)
         @Override
         protected synchronized InterceptibleThread newThread(ThreadGroup threadGroup, Runnable runnable, String name)
         {
-            InterceptibleThread thread = new InterceptibleThread(threadGroup, runnable, name, extraToStringInfo, onTermination, parent.interceptorOfGlobalMethods, time);
+            // Can not use NamedThreadFactory.globalPrefix() as this method runs in the App class loader and not the Instance class loader; the ThreadGroup's name can act as a proxy for this.
+            String threadName = threadGroup.getName() + '_' + name;
```
Comment (dcapwell) [inline test/simulator/main/org/apache/cassandra/simulator/systems/InterceptibleThreadFactory.java:62]: so with Simulator doing threads itself it was missing the node prefix, which made the logs far harder to understand... so i keep the same semantics as normal jvm-dtest.

File: test/simulator/main/org/apache/cassandra/simulator/systems/SimulatedTime.java [hunk 17]
```diff
@@ -274,7 +274,7 @@ public long nanoTime()
                 nextDrift = nanosDriftSupplier.get(random);
                 from = global;
                 to = global + Math.max(baseDrift, nextDrift);
-                diffPerGlobal = (nextDrift - baseDrift) / (double)(to - from);
+                diffPerGlobal = to == from ? 1 : (nextDrift - baseDrift) / (double)(to - from);
```
Comment (dcapwell) [inline test/simulator/main/org/apache/cassandra/simulator/systems/SimulatedTime.java:277]: when i monkeyed with controlling time i hit this case... i wanted to test the case where the full test ran unrealistically fast (each operation only took 1ns)...

File: src/java/org/apache/cassandra/batchlog/BatchlogManager.java [hunk 18]
```diff
@@ -431,7 +431,7 @@ public boolean replay(RateLimiter rateLimiter, Set<UUID> hintedNodes) throws IOE
             if (accordMutations != null)
             {
                 accordTxnStart = accordTxnStart.withStartedAt(Clock.Global.nanoTime());
-                accordResult = accordMutations != null ? mutateWithAccordAsync(cm, accordMutations, null, accordTxnStart) : null;
+                accordResult = accordMutations != null ? mutateWithAccordAsync(cm, accordMutations, null, accordTxnStart, true) : null;
```
Comment (aweisberg) [inline src/java/org/apache/cassandra/batchlog/BatchlogManager.java]: Can you make preserve timestamps an enum so that we can then document the enum values so it will be easy to discover anywhere we use the enum what it's all about?

File: src/java/org/apache/cassandra/cql3/statements/BatchStatement.java (hunk 1 above)
Comment (aweisberg) [reply src/java/org/apache/cassandra/cql3/statements/BatchStatement.java]: I don't think it will ever be that useful? We pass in whether to preserve timestamps to `mutateWithTriggers` and I think that is enough. We don't need to serialize it with the mutation because it's only needed for batch log and hints where it is flipped to true anyways by necessity.

Anything that fully addresses it will need a per statement granularity preserving timestamp.

Anything that partially addresses it can use the boolean passed `mutateWithTriggers` to signal an error in cases where the timestamp wouldn't be preserved.

File: src/java/org/apache/cassandra/db/Mutation.java [hunk 19]
```diff
@@ -86,6 +87,7 @@ public class Mutation implements IMutation, Supplier<Mutation>
     volatile long viewLockAcquireStart;
 
     private final boolean cdcEnabled;
+    private final boolean userTimestamp;
```
Comment (aweisberg) [inline src/java/org/apache/cassandra/db/Mutation.java]: I don't think this flag makes sense or has utility? Only the coordinator needs to make the call about whether to preserve the timestamp. It doesn't need to be serialized and carried with the mutation.

Batchlog and hints always preserve mutations so they don't need it serialized.

The granularity isn't enough to do anything useful at the coordinator and `mutateWithTriggers` already has this information passed in as a parameter. I really don't think it's worth the noise in `Mutation` if it's never going to be usable for anything.

File: src/java/org/apache/cassandra/db/Mutation.java (hunk 3 above)
Comment (aweisberg) [reply src/java/org/apache/cassandra/db/Mutation.java]: Using a vint is fine, but `write` with a 32-bit int writes 1 byte not 4. With versioning of serialization we don't really need to burn instructions on a vint unless we want to add flags without bumping the version?

File: src/java/org/apache/cassandra/service/StorageProxy.java [hunk 20]
```diff
@@ -1248,20 +1249,22 @@ public static void mutateWithTriggers(List<? extends IMutation> mutations,
...
                 List<? extends IMutation> accordMutations = splitMutations.accordMutations();
-                IAccordResult<TxnResult> accordResult = accordMutations != null ? mutateWithAccordAsync(cm, accordMutations, consistencyLevel, requestTime) : null;
                 List<? extends IMutation> normalMutations = splitMutations.normalMutations();
+                if (!preserveTimestamps && normalMutations != null)
```
Comment (aweisberg) [inline src/java/org/apache/cassandra/service/StorageProxy.java]: Add a comment explaining that if there was ever any attempt to apply part of the mutation using the eventually consistent path then we need to continue to use the timestamp used by the eventually consistent path to not end up with multiple timestamps, but if it only ever used the transactional path then we can use the transactional timestamp to get linearizability

File: src/java/org/apache/cassandra/service/accord/txn/TxnWrite.java [hunk 21]
```diff
@@ -86,23 +88,26 @@ public class TxnWrite extends AbstractKeySorted<TxnWrite.Update> implements Writ
...
     public static class Update extends AbstractParameterisedVersionedSerialized<PartitionUpdate, TableMetadatas>
     {
-        private static final long EMPTY_SIZE = ObjectSizes.measure(new Update(null, 0, ByteBufferUtil.EMPTY_BYTE_BUFFER));
+        private static final long EMPTY_SIZE = ObjectSizes.measure(new Update(null, 0, ByteBufferUtil.EMPTY_BYTE_BUFFER, false));
         public final PartitionKey key;
         public final int index;
+        public final boolean userTimestamp;
 
-        public Update(PartitionKey key, int index, PartitionUpdate update, TableMetadatas tables)
+        public Update(PartitionKey key, int index, PartitionUpdate update, TableMetadatas tables, boolean userTimestamp)
```
Comment (aweisberg) [inline src/java/org/apache/cassandra/service/accord/txn/TxnWrite.java]: I don't think this is the right granularity for `userTimestamp` because it's a whole partition update which is multiple rows some of which might preserve timestamp and some of which might not.

We already capture what to do with timestamps with `preserveTimestamps` for the whole partition update so it's redundant to add `userTimestamp`?

File: src/java/org/apache/cassandra/service/consensus/migration/ConsensusMigrationMutationHelper.java [hunk 22]
```diff
@@ -262,17 +266,18 @@ public static IAccordResult<TxnResult> mutateWithAccordAsync(ClusterMetadata cm,
         long minEpoch = Epoch.EMPTY.getEpoch();
         for (IMutation mutation : mutations)
         {
+            boolean userTimestamp = mutation instanceof Mutation && ((Mutation) mutation).isUserTimestamp();
```
Comment (aweisberg) [inline src/java/org/apache/cassandra/service/consensus/migration/ConsensusMigrationMutationHelper.java]: Again, this isn't a useful granularity. In practice mutation is not a grouping upon which `userTimestamp` is set its per row?

We don't need this intermediate step that only works with some CQL but not others and adds a lot of noise and needs to be replaced anyways.

Just the flag to `mutateWithTriggers` is enough to either error out on user timestamps, or preserve them. Partially linearizable batches is just wonky and not really worth what this amount of code achieve that needs to be replaced anyways?

File: test/distributed/org/apache/cassandra/distributed/test/cql3/StatefulASTBase.java [hunk 23]
```diff
@@ -206,10 +211,55 @@ protected <S extends BaseState> Property.StatefulSuccess<S, Void> onSuccess(Logg
...
+        });
+    }
+
     protected static <S extends CommonState> Property.Command<S, Void, ?> insert(RandomSource rs, S state)
     {
         int timestamp = ++state.operations;
-        Mutation mutation = state.mutationGen().next(rs).withTimestamp(timestamp);
+        Mutation original = state.mutationGen().next(rs);
+        Mutation mutation = state.allowUsingTimestamp()
+                            ? original.withTimestamp(timestamp)
```
Comment (aweisberg) [inline test/distributed/org/apache/cassandra/distributed/test/cql3/StatefulASTBase.java:261]: This is changing the timestamp of the entire mutation and not specific rows in it which is why the implementation driven by this test satisfies the test, but doesn't actually work correctly for real world unlogged single partition batches which can have different rows with different timestamps.

File: test/simulator/main/org/apache/cassandra/simulator/systems/SimulatedTime.java (hunk 17 above)
Comment (aweisberg) [reply test/simulator/main/org/apache/cassandra/simulator/systems/SimulatedTime.java:277]: I spent 5 minutes looking at this and didn't grok what `diffPerGlobal` is doing.

**Review by aweisberg — CHANGES_REQUESTED** [review]
My initial reaction is that this approach of having per mutation flags on whether `USING TIMESTAMP` is set doesn't make much sense. It generates a lot of flags passed around and serialized while not accomplishing the goal of supporting `USING TIMESTAMP` and also not safely blocking the scenarios we don't support.

Like `attrs.usingTimestamp` inside `BatchStatement` doesn't make sense because when could that ever be set? It should only be set on `INSERT`/`UPDATE`? Maybe I am missing how that is used/set.

Ignoring `TransactionStatement` for the moment which I think can fully block `USING TIMESTAMP` unless we implement it correctly at the statement level.

If we do something at this stage that doesn't support everything then I think we should make it so that you get an error if you try to mixing `USING TIMESTAMP` statements with not `USING TIMESTAMP` and head down the transactional path.

If we want to do this right then I think we should find a way to finagle the information we need into `TxnUpdate` and I would be fine with passing in sidecar data structure that maps statements to their timestamp preservation requirements. This would be empty if `USING TIMESTAMP` is not in use so it's low overhead and pay for what you use.

This also avoids having to add flags to `BTreeRow`. You can't actually use cell level flags unless we are sure that there will always be cells. I think this is actually true because you can't insert without providing the primary key which means at least those cells are present. If we set a flag on one of them we could then trivially know how to handle timestamps for all cells in the row.

File: src/java/org/apache/cassandra/db/Mutation.java (hunk 3 above)
Comment (dcapwell) [reply src/java/org/apache/cassandra/db/Mutation.java]: you right right, i got confused thinking this was `write int` but its `write byte`.

> With versioning of serialization we don't really need to burn instructions on a vint unless we want to add flags without bumping the version?

We get 8 flags before we care and this patch gets us to 2; so we really don't need it atm.

Ill revert back to `write

File: test/simulator/main/org/apache/cassandra/simulator/systems/SimulatedTime.java (hunk 17 above)
Comment (dcapwell) [reply test/simulator/main/org/apache/cassandra/simulator/systems/SimulatedTime.java:277]: i think its trying to make sure each instance clock isn't 100% in-sync so that they are different by some amount.  I didn't go into too much detail as my test is single node and i wanted to focus on operations that take 1ns...  I reverted the 1ns testing as that wasn't the issue i was facing.

Comment (belliottsmith) [reply test/simulator/main/org/apache/cassandra/simulator/systems/SimulatedTime.java:277]: I think it is to ensure monotonicity when the drift is changed, since we use this clock for both monotonic and non-monotonic timestamps. That is, we apply the new drift as of some future time, and we smear the change in drift over the time period.

File: test/simulator/main/org/apache/cassandra/simulator/systems/InterceptibleThreadFactory.java (hunk 16 above)
Comment (belliottsmith) [reply test/simulator/main/org/apache/cassandra/simulator/systems/InterceptibleThreadFactory.java:62]: Is there no way to have the logs include the thread group name, rather than replicating it?

File: test/simulator/main/org/apache/cassandra/simulator/Actions.java (hunk 12 above)
Comment (belliottsmith) [reply test/simulator/main/org/apache/cassandra/simulator/Actions.java]: Could you explain better the rationale for using an EmptyAction here, rather than no action at all?

File: test/simulator/main/org/apache/cassandra/simulator/ActionSchedule.java (hunk 11 above)
Comment (belliottsmith) [reply test/simulator/main/org/apache/cassandra/simulator/ActionSchedule.java:77]: I appreciate the docs improvement, but (as with much AI content) there's a lot of redundant/repetitious information here. I'm not sure the "Use cases" headings offer much value to the reader.

File: test/simulator/main/org/apache/cassandra/simulator/ActionSchedule.java [hunk 24]
```diff
@@ -73,7 +73,65 @@ public class ActionSchedule implements CloseableIterator<Object>, LongConsumer
...
+         * <ul>
+         *     <li>Tests that should terminate on whichever condition happens first</li>
+         *     <li>Flexible scenarios requiring either completion or timeout</li>
+         *     <li>Testing resilience against both completion and time constraints</li>
+         * </ul>
+         */
+        TIME_AND_STREAM_LIMITED,
+        /**
+         * Definition: Does not allow stream actions at all
+         * Behavior: Throws IllegalStateException if stream actions are added; no special handling for daemon tasks
```
Comment (belliottsmith) [inline test/simulator/main/org/apache/cassandra/simulator/ActionSchedule.java]: Is the relevant fact that we throw ISE, or that this mode is intended to be used with a FINITE provision of actions, whereas a STREAM is intended to provide an unlimited supply of actions? ISE is simply verification that this is sensible usage.

File: test/simulator/main/org/apache/cassandra/simulator/ActionSchedule.java [hunk 25]
```diff
@@ -73,7 +73,65 @@ public class ActionSchedule implements CloseableIterator<Object>, LongConsumer
...
+         * Use cases:
+         * <ul>
+         *     <li>Testing deterministic scenarios with a fixed set of actions</li>
+         *     <li>Avoiding recurring or streaming operations</li>
+         *     <li>Unit tests with controlled inputs</li>
+         * </ul>
+         */
+        FINITE,
+        /**
+         * Definition: Focused on daemon task management without time/stream constraints
```
Comment (belliottsmith) [inline test/simulator/main/org/apache/cassandra/simulator/ActionSchedule.java]: This is strange framing. The way daemon tasks are managed is a side effect, not a goal. We can perhaps rename this, but the only distinction with FINITE is that we expect daemon tasks to outlive the work in this phase, but the work in the phase is expected to be finite otherwise. So we manage the daemon tasks in waves so that we can know when the "real" work is actually completed.

File: test/simulator/main/org/apache/cassandra/simulator/ClusterSimulation.java (hunk 13 above)
Comment (belliottsmith) [reply test/simulator/main/org/apache/cassandra/simulator/ClusterSimulation.java:152]: fwiw, I think Ariel's work has landed so we're free to explore simulator issues :)

File: test/simulator/main/org/apache/cassandra/simulator/systems/SimulatedTime.java (hunk 17 above)
Comment (belliottsmith) [reply test/simulator/main/org/apache/cassandra/simulator/systems/SimulatedTime.java:277]: Probably we should simulate non-monotonic system time changes

File: test/simulator/main/org/apache/cassandra/simulator/Actions.java (hunk 12 above)
Comment (dcapwell) [reply test/simulator/main/org/apache/cassandra/simulator/Actions.java]: > rather than no action at all?
you mean return `null`?

The use case is a action is selected in the AST logic, but it can't apply given the current state of the model (can't select a known partition if there are no partitions), so need a way to detect this and retry with a different action

File: test/simulator/main/org/apache/cassandra/simulator/ClusterSimulation.java (hunk 13 above)
Comment (dcapwell) [reply test/simulator/main/org/apache/cassandra/simulator/ClusterSimulation.java:152]: i can look to add another multi node test

File: test/simulator/main/org/apache/cassandra/simulator/ActionSchedule.java (hunk 11 above)
Comment (dcapwell) [reply test/simulator/main/org/apache/cassandra/simulator/ActionSchedule.java:77]: i had found some value as i was struggling to understand why my `infiniteStream` based test was not hitting the max steps i had encoded in the test... learned that I wanted a `stream` instead

I removed the `Use cases` section

File: test/simulator/main/org/apache/cassandra/simulator/systems/InterceptibleThreadFactory.java (hunk 16 above)
Comment (dcapwell) [reply test/simulator/main/org/apache/cassandra/simulator/systems/InterceptibleThreadFactory.java:62]: looking and the best i see is creating a annoying plugin for logback that has to be linked in the config as well, so would need a java plugin we write + config changes.

having the thread name prefixed with the node is what we do in jvm-dtest and you can find tests leverage this behavior.  Its also be really great when taking stack traces when things go wrong, or collecting stack dumps.  For example we see that simulator keeps failing in CI due to leaked threads... but from which nodes?  Does thread X relate to thread Y or is it from a different node?

File: test/simulator/main/org/apache/cassandra/simulator/ActionSchedule.java (hunk 24 above)
Comment (dcapwell) [reply test/simulator/main/org/apache/cassandra/simulator/ActionSchedule.java]: i can reword, but do think its important to call out this property; the details of what exception isn't needed here.

I went with

> Behavior: Processes a finite set of actions, if any action is added that has stream or daemon like behavior then the simulation will be rejected.

File: test/simulator/main/org/apache/cassandra/simulator/ActionSchedule.java (hunk 25 above)
Comment (dcapwell) [reply test/simulator/main/org/apache/cassandra/simulator/ActionSchedule.java]: tried to reword this

File: test/simulator/main/org/apache/cassandra/simulator/Actions.java (hunk 12 above)
Comment (belliottsmith) [reply test/simulator/main/org/apache/cassandra/simulator/Actions.java]: If this is a sentinel action for that specific use case, maybe move it there and give it a suitable name for the purpose? Doesn't seem to make much sense in isolation.

File: test/simulator/main/org/apache/cassandra/simulator/ActionSchedule.java (hunk 24 above)
Comment (belliottsmith) [reply test/simulator/main/org/apache/cassandra/simulator/ActionSchedule.java]: I think that's incorrect? We don't reject daemon like behaviour, I think we just expect it to be finite (and may not terminate if that doesn't hold)

File: doc/modules/cassandra/pages/managing/operating/onboarding-to-accord.adoc [hunk 26]
```diff
@@ -279,14 +279,29 @@ Supported read consistency levels are `ONE`, `QUORUM`, `SERIAL`, and
...
 
 `USING TIMESTAMP` is allowed and the application of the operations will
 occur in a linearizable order, but from the perspective of a reader the
 merged result may not appear linearizable.
 
+For BATCH operations, timestamp handling is more complex:
+- If the batch uses `USING TIMESTAMP`, the user timestamp will be used
+- If all mutations in a BATCH use `USING TIMESTAMP`, the user timestamp will be used
+- If not using `USING TIMESTAMP` and all partition keys are on Accord, the Accord timestamp is used
+- If the BATCH has some partitions on Accord and others not on Accord, the server timestamp will be used (unsafe for multi-table batches where one table is not migrating to Accord)
```
Comment (aweisberg) [inline doc/modules/cassandra/pages/managing/operating/onboarding-to-accord.adoc]: Unsafe is a little non-specific. The writes to the Accord table will not be linearizable.

File: doc/modules/cassandra/pages/managing/operating/onboarding-to-accord.adoc [hunk 27]
```diff
@@ -279,14 +279,29 @@ Supported read consistency levels are `ONE`, `QUORUM`, `SERIAL`, and
...
-non-SERIAL operations are not linearizable even when executed on Accord
-because Accord will continue to write data using the coordinator
-generated timestamp not the transaction's timestamp.
+non-SERIAL operations are linearizable by default when migrated to Accord,
+but during migration they might not be linearizable due to timestamp handling
+changes. During migration, normal CQL operations will use the Accord 
+timestamp once a range starts migration, but will fall back to server 
+timestamp when migrating away from Accord.
+
+Hints will use server timestamp and they only come into play during
```
Comment (aweisberg) [inline doc/modules/cassandra/pages/managing/operating/onboarding-to-accord.adoc]: And batch log replay.

Also might be worth explaining that if only Accord tables are used the batch will not be written to the batch log, but if it's a mix of Accord and non-Accord tables then the batch will be written to the batch log.

File: doc/modules/cassandra/pages/architecture/cql-on-accord.adoc [hunk 28]
```diff
@@ -253,10 +253,24 @@ at each replica when `TxnWrite.apply` is called.
...
+`USING TIMESTAMP` is supported and will cause the data to use the user
+timestamp instead of the Accord one, though this breaks linearizability
+and should be avoided when possible.
+
+For BATCH operations, timestamp handling is more complex: if the batch
+uses `USING TIMESTAMP`, the user timestamp will be used. If all mutations
+in a BATCH use `USING TIMESTAMP`, the user timestamp will be used. If not
+using `USING TIMESTAMP` and all partition keys are on Accord, the Accord
+timestamp is used. If the BATCH has some partitions on Accord and others
+not on Accord, the server timestamp will be used (unsafe for multi-table
```
Comment (aweisberg) [inline doc/modules/cassandra/pages/architecture/cql-on-accord.adoc]: Instead of unsafe, not linearizable for the writes to Accord tables.

**Review by aweisberg — APPROVED** [review]
Minor doc nits. Benedict had an unresolved ask with `EmptyAction` in simulator.

File: test/simulator/main/org/apache/cassandra/simulator/Actions.java (hunk 12 above)
Comment (dcapwell) [reply test/simulator/main/org/apache/cassandra/simulator/Actions.java]: ill try to replace with with `null` as the logic is checked before sending to simulator

File: test/simulator/main/org/apache/cassandra/simulator/ActionSchedule.java (hunk 24 above)
Comment (dcapwell) [reply test/simulator/main/org/apache/cassandra/simulator/ActionSchedule.java]:
```
case FINITE:
                if (action.is(STREAM)) throw new IllegalStateException();
```

The only check for finite is that it isn't a stream; updated the docs

File: test/simulator/main/org/apache/cassandra/simulator/Actions.java (hunk 12 above)
Comment (dcapwell) [reply test/simulator/main/org/apache/cassandra/simulator/Actions.java]: switched to `null`

File: test/distributed/org/apache/cassandra/distributed/test/cql3/StatefulASTBase.java (hunk 23 above)
Comment (dcapwell) [reply test/distributed/org/apache/cassandra/distributed/test/cql3/StatefulASTBase.java:261]: sorry thought this comment was for main code and not test.

`withTimestamp` is currently only implemented in INSERT / UPDATE / DELETE.  Batch support never got merged (see https://issues.apache.org/jira/browse/CASSANDRA-20503) as there were annoying edge cases with unexpected behavior and things were ready to start testing `BEGIN TRANSACTION`.

Even if batch was added, it would just add the timestamp to the top level batch and wouldn't do the "mix" that this patch guards against

Comment (dcapwell) [reply test/distributed/org/apache/cassandra/distributed/test/cql3/StatefulASTBase.java:261]: > but doesn't actually work correctly for real world unlogged single partition batches which can have different rows with different timestamps.

The test isn't trying to cover that, the whole `USING TIMESTAMP` logic is to make sure that the test is strict serializable else the model fails.  When we run on accord the idea is that we no longer need to use `USING TIMESTAMP` as Accord makes sure this is safe.

**Comment by dcapwell** [comment]
merged

```
commit 6adac494aa534696e552f28450f4372faceac779
Author: David Capwell <dcapwell@apache.org>
Date:   Wed Jul 30 10:03:47 2025 -0700

    When regulars CQL mutations run on Accord use the txn timestamp rather than server timestamp

    patch by David Capwell; reviewed by Ariel Weisberg for CASSANDRA-20744
```
